src/include/migraphx/run_loop.hpp

```diff
@@ -0,0 +1,120 @@
+#ifndef MIGRAPHX_GUARD_RTGLIB_RUN_LOOP_HPP
+#define MIGRAPHX_GUARD_RTGLIB_RUN_LOOP_HPP
+
+#include <string>
+#include <migraphx/instruction_ref.hpp>
+#include <migraphx/shape.hpp>
+#include <migraphx/argument.hpp>
+#include <migraphx/context.hpp>
+#include <migraphx/module.hpp>
+#include <migraphx/config.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+static inline std::pair<int, bool> get_name_index(const std::string& name,
+                                                  const std::string& param_prefix)
+{
+    auto loc = name.find(param_prefix);
+    if(loc != std::string::npos)
+    {
+        int index = std::stoi(name.substr(loc + param_prefix.size()));
+        return {index, true};
+    }
+
+    std::string out_prefix = "#output_";
+    loc                    = name.find(out_prefix);
+    if(loc != std::string::npos)
+    {
+        int index = std::stoi(name.substr(loc + out_prefix.size()));
+        return {index, false};
+    }
+
+    return {-1, false};
+}
+
+template <class LoopModel, class T>
+argument run_loop(const LoopModel& model,
+                  T& ctx,
+                  std::vector<argument> args,
+                  const std::vector<module_ref>& mods,
+                  const std::function<std::vector<argument>(
+                      module_ref&, const std::unordered_map<std::string, argument>&)>& run)
+{
+    std::vector<std::vector<argument>> results;
+    // process argu lists
+    auto iter_num = args.at(0).at<int64_t>();
+    auto cond     = args.at(2).at<bool>();
+
+    args.erase(args.begin() + 2);
+    args.erase(args.begin());
+
+    auto input_num = args.size() - 2;
+    auto dep_num   = input_num - 2;
+
+    module_ref mod           = mods.at(0);
+    auto param_name_shapes   = mod->get_parameter_shapes();
+    std::string param_prefix = "#" + mod->name() + "_in_";
```

Comment:
I understand why we add a prefix when doing the onnx parser so each parameter can be orderded correctly. However, for `run_loop` it should just sort the parameter names and the index into this sorted vector is parameter that should be used.

If we(or a user) build a loop operator manually then we might not name this operators exactly the same. The important part is that each arg is associated with each parameter as its sorted.

Reply:
The problem is, some input parameters can be deleted when we compile a module. If some parameters are removed, even we sort the parameters, we do not know the relationship between input arguments and corresponding parameters since some parameters could have been removed. Maybe we should not remove input parameters in dead_code_elimination, then we can do what you proposed above.